### PR TITLE
requireMatchingFunctionName: fix critical bug and add tests

### LINKED ISSUE #1601: requireMatchingFunctionName error with predefined variables
If you have defined a variable name before, then assign it to an (anonymous) function, requireMatchingFunctionName will crash. Example code:

```javascript
var abc = function() {
  var myFunction;

  myFunction = function(name) {
    console.log('hello', name);
  };
};
```
The error message:
>Error running rule requireMatchingFunctionName: This is an issue with JSCS and not your codebase.
Please file an issue (with the stack trace below) at: https://github.com/jscs-dev/node-jscs/issues/new
TypeError: Cannot read property 'name' of undefined
    at checkForMember (file:///C:/Users/Famlam/Desktop/RedundancyCheck/test%20suite/jscs.js:10499:38)
    at file:///C:/Users/Famlam/Desktop/RedundancyCheck/test%20suite/jscs.js:10473:21
    at Array.forEach (native)
    at Object.3.JsFile.iterateNodesByType (file:///C:/Users/Famlam/Desktop/RedundancyCheck/test%20suite/jscs.js:1755:42)
    at Object.83.module.exports.check (file:///C:/Users/Famlam/Desktop/RedundancyCheck/test%20suite/jscs.js:10468:14)
    at Object.<anonymous> (file:///C:/Users/Famlam/Desktop/RedundancyCheck/test%20suite/jscs.js:17118:22)
    at Array.forEach (native)
    at Object.139.StringChecker._checkJsFile (file:///C:/Users/Famlam/Desktop/RedundancyCheck/test%20suite/jscs.js:17114:31)
    at Object.139.StringChecker.checkString (file:///C:/Users/Famlam/Desktop/RedundancyCheck/test%20suite/jscs.js:17059:14)
    at Object.<anonymous> (file:///C:/Users/Famlam/Desktop/RedundancyCheck/test%20suite/index.html?module=Other:5008:32) at input :

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/25260334-requirematchingfunctionname-error-with-predefined-variables?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by zxqfox**
Fixes #1601

### CODE REVIEW

File: lib/rules/require-matching-function-name.js [hunk 1]
```diff
@@ -96,11 +96,13 @@ module.exports.prototype = {
 };
 
 function checkForMember(assignment, skip, errors) {
-    if (skip(assignment.left.property.name, assignment.right)) {
+    var _left = assignment.left;
+    var _name = _left.property ? (_left.property.name || _left.property.value) : _left.name;
```
Comment (mrjoelkemp) [inline lib/rules/require-matching-function-name.js]: Non-blocker: For readability, I find it helpful for maintenance when an example code is supplied that clarifies an assignment with a left property that doesn't have a name but has a value. Similarly, what's an assignment that doesn't have a property?

A contributor would have to guess or hope that coverage is 100% and figure out which tests correspond to which part of the condition.

**Comment by mrjoelkemp** [comment]
LGTM aside from the JSHint complaints.

File: lib/rules/require-matching-function-name.js (hunk 1 above)
Comment (zxqfox) [inline lib/rules/require-matching-function-name.js]: There are three options of assignments:
```
var simpleVar = function () {};
obj.propertyName = function () {};
obj['propertyValue'] = function () {};
```
And _left property only exists for the second and the third line.

I'll add here comments.

Comment (mrjoelkemp) [inline lib/rules/require-matching-function-name.js]: Comments would be great. Thanks @zxqfox!

**Comment by zxqfox** [comment]
Would be nice to have another review.

**Comment by mikesherov** [comment]
:+1:

File: test/specs/rules/require-matching-function-name.js [hunk 2]
```diff
@@ -45,10 +65,30 @@ describe('rules/require-matching-function-name', function() {
...
+
+        it('should NOT report function assignation to an expression', function() {
+            assertNoErrors('var test = {}; test["x" + 2] = function bar(){};');
+        });
+
+        it('should NOT report function name match on ES6 destructuring', function() {
+            assertNoErrors('let [ bar ] = [ function bar(){} ];');
+        });
+
+        it.skip('should report function name mismatch on ES6 destructuring', function() {
```
Comment (mikesherov) [inline test/specs/rules/require-matching-function-name.js]: On purpose?

Comment (zxqfox) [inline test/specs/rules/require-matching-function-name.js]: For a second I though we need to check names for ES6-destructuring. Feels like need to drop this test.

**Pull request closed without merging**